File: assets/src/array-test.py
# coding: utf-8
# さまざまな配列と Iterator パターン

# タプル(tuple): 作成した後は要素の追加や削除を行えない配列

import datetime

def get_today_tuple():
  today = datetime.datetime.today()             # 現在の日にちを取得
  tuple1 = (today.year, today.month, today.day)  # タプルによる定義: (年, 月, 日)
  return tuple1

today = get_today_tuple()
print(today)
for i in today:  # 外部イテレータ: オブジェクトの外から要素を走査して取り出す。典型的には for 文を用いる。
  print(i)

# リスト(list): 作成した後も要素の追加や削除を行える配列

list = ['array', '-', 'test'] # リストの定義
print(list)
for i in list:  # 外部イテレータ
  print(i)

list.append('.')  # 要素の追加
list.append('py')
print(list)

list.insert(0, 'python') # 添字を指定しての要素の追加
print(list)

list.remove('-') # 要素の削除 先頭から探して最初に見つかった要素を削除する
list.remove('.')
print(list)

print(list.pop(0)) # 添字を指定しての要素の削除: 先頭の要素の削除
print(list)
print(list.pop())  # 添字を指定しての要素の削除: 末尾の要素の削除
print(list)

print(list.index('test')) # 要素の検索 先頭から探して最初に見つかった要素の添字を返す

print(list.count('array')) # 指定した値に一致する要素の数を数える

# 辞書 または ディクショナリ(dictionary): キーと値の対応関係を定義できる。他の言語では，連想配列，マップ，ハッシュともいう。
def get_today_dic():
  today = datetime.datetime.today()             # 現在の日にちを取得
  dic = {     # 辞書による定義: {年, 月, 日}
    'year': today.year,
    'month': today.month,
    'day': today.day
  }
  return dic

today = get_today_dic()
print(today)
print(today['year'])  # 辞書による参照: 配列っぽく書いた例
print(today.get('month')) # 辞書による参照: get を使った例
print(today.get('day', 'Not found')) # 辞書による参照: 見つからなかった場合の値を定義する例 
print(today.get('date', 'Not found')) # 辞書による参照: 見つからなかった場合の値を定義する例 

for i in today: # 外部イテレータ
  print(i)
  print(today[i])

# 集合 または セット(set): 値が重複できない配列。順番も保持されない。
set1 = set() # 空集合: set1 = {} としてしまうと辞書になってしまう
print(set1)
set1 = {'array', '-', 'test'} # こう書くと辞書ではなく集合
print(set1) # 順番は維持されない

for i in set1: # 外部イテレータ
  print(i)

set1.add('python') # 1要素の追加
print(set1)
set1.update({'.', 'py'}) # 他の集合からの要素の追加 (同じようにリストやタプルからも追加できる)
print(set1)

set1.remove('.') # 要素の削除: 指定した要素が存在しない場合はエラーになる
set1.discard('py') # 要素の削除: 指定した要素が存在しなくてもエラーにならない
set1.discard('py') # 存在しない要素を削除する例。remove に書き換えると KeyError となる
print(set1)

set2 = frozenset(set1) # 要素を変更できない集合を取得する
print(set2)
# set2.add('py') # frozenset に要素を追加すると AttributeError となる
# set2.update({'.'}) # frozenset に要素を追加すると AttributeError となる
# set2.remove('-') # frozenset に要素を削除すると AttributeError となる
# set2.discard('-') # frozenset に要素を削除すると AttributeError となる

# 内部イテレータの例: map, filter, reduce

items = [1, 2, 3]

def square(n): # 二乗する関数
  return n * n

# map: 各要素について処理を行う
print(map(square, items)) # map 関数でリスト items の各要素を二乗したリストを返す
print(map(lambda n: n * n, items)) # lamdbaで無名関数を定義することができる。先ほどと同じことが関数を別に定義しなくてもできる
print([x * x for x in items]) # 同じことを外部イテレータを使ってこういう風にも書ける

print(filter(lambda n:n % 2 != 0, items)) # filter 関数で奇数のみを取り出す
print([x for x in items if x % 2 != 0]) # 同じことを外部イテレータを使ってこういう風にも書ける

# reduce: 全部を1つにまとめる

def add(a, b):
  return a + b

print(reduce(add, items)) # reduce 関数で総和を求める: add(add(1, 2), 3)
print(reduce(lambda a, b: a + b, items)) # 同じことを無名関数ではこう書く

# map, filter, reduce を組み合わせると，配列に対する様々な操作を簡潔に書くことができる

print(reduce(add, map(square, items))) # 二乗和の計算
print(reduce(add, filter(lambda n:n % 2 != 0, items))) # 要素から奇数のみ取り出して総和を求める
print(reduce(add, map(square, filter(lambda n:n % 2 != 0, items)))) # 要素から奇数のみ取り出して二乗和を求める

File: assets/src/array-test.cpp
// さまざまな配列と Iterator パターン

#include <time.h> // 年月日の取得用
#include <array> // 配列(array): 固定長配列，要素の追加や削除を行えない
#include <vector> // 可変長配列 または 動的配列 (vector): 要素の追加や削除を行える
#include <map> // マップ(map): キーと値の対応関係を定義できる。他の言語では辞書，ディクショナリ，ハッシュともいう
#include <set> // 集合またはセット(set): 値が重複できない。順番も保持されない。複数の値を持つ集合は multiset
#include <iterator> // イテレータ
#include <algorithm> // transform, copy_if
#include <numeric> // accumelate
#include <iostream>
#include <string>

using namespace std;

int square(int n) {
  return n * n;
}

int add(int a, int b) {
  return a + b;
}

int isOdd(int n) {
  return n % 2 != 0;
}

int main()
{
  // 固定長配列
  array<int, 3> today1; // 固定長配列の定義
  time_t now_t = time(nullptr); // 現在時刻の取得
  struct tm *now = localtime(&now_t); // 変換
  today1[0] = now->tm_year + 1900; // 年　代入は普通の配列と同じようにできる
  today1[1] = now->tm_mon + 1; // 月
  today1[2] = now->tm_mday; // 日

  cout << today1[0] << endl; // 参照も普通の配列と同じようにできる
  cout << today1[1] << endl;
  cout << today1[2] << endl;

  for(array<int, 3>::iterator i = begin(today1); // 外部イテレータの古い記法
    i != end(today1);
    i++) {
    cout << *i << endl; // * がつく点に注意
  }

  for(auto& i: today1) { // 外部イテレータの C++11 にしたがった新しい記法
    cout << i << endl; // * がつかない点に注意
  }

  // 可変長配列
  vector<string> v1; // 空の可変長配列 v1 の定義
  string org_v[] = {"array", "-", "test"};
  vector<string> v2(org_v, end(org_v)); // 配列を元に v2 を定義する古い記法
  vector<string> v3{"array", "-", "test"}; // 配列を元に v3 を定義する C++11 にしたがった新しい記法

  for(vector<string>::iterator i = begin(v2); // 外部イテレータの古い記法
    i != end(v2);
    i++) {
    cout << *i << endl;
  }

  for(auto& i: v3) { // 外部イテレータの C++11 にしたがった新しい記法
    cout << i << endl;
  }

  // マップ
  map<string, int> today2; // マップの定義
  today2["year"] = now->tm_year + 1900; // 年
  today2["month"] = now->tm_mon + 1; // 月
  today2["day"] = now->tm_mday; // 日

  cout << today2["year"] << endl; // 参照
  cout << today2.at("month") << endl; // 参照は at を用いても良い。	

  const map<string, int> today3(today2); // 変更不能なマップの定義
  cout << today3.at("day") << endl; // 変更不能(const)で定義しているときは at で定義しているときは at で参照する

  // 存在するかどうか判定する
  if(today2.count("date") == 0) {
    cout << "\"date\" is not found." << endl;
  } else {
    cout << "\"date\" is found." << endl;
  }

  // cout << today2.at("date") << endl; // 存在しないキーの値を at で参照すると key not found エラーになる

  cout << today2["date"] << endl; // 要注意: 存在しないキーの値を参照すると 0 になる

  // 要注意: 一度 [] で存在しない要素を参照すると要素が存在することになってしまう
  if(today2.count("date") == 0) {
    cout << "\"date\" is not found." << endl;
  } else {
    cout << "\"date\" is found." << endl;
  }

  cout << today2.at("date") << endl; // 存在していることになっているので key not found エラーにならない

  today2.erase("date"); // 要素の削除 存在しない場合は無視される

  for(map<string, int>::iterator i = today2.begin();	// 外部イテレータの古い記法
    i != today2.end();
    i++) {
    cout << "key = " << i->first // -> で参照することに注意
      << ", value = " << i->second << endl;
	}

  for(auto& i: today2) { // 外部イテレータの C++11 にしたがった新しい記法
    cout << "key = " << i.first // . で参照することに注意
      << ", value = " << i.second << endl;
  }

  // 集合
  set<string> set1; // 空集合
  set<string> set2(org_v, end(org_v)); // 集合を配列で初期化する
  set<string> set3{"array", "-", "test"}; // 集合の初期化のC++11にしたがった新しい記法

  for(set<string>::iterator i = set2.begin(); // 外部イテレータの古い記法
    i != set2.end();
    i++) {
    cout << *i << endl; // * がつくことに注意
  }

  for(auto& i: set3) { // 外部イテレータの C++11 にしたがった新しい記法
    cout << i << endl; // * がつかないことに注意
  }

  set2.insert("python"); // 要素の追加

  for(auto& i: set2) {
    cout << i << endl;
  }

  auto r = set2.insert("python"); // 要素の追加に戻り値もある
  if(r.second) {
    cout << "python is added newly." << endl;
  } else {
    cout << "python has already been added." << endl;
  }

  auto c = set2.erase("python"); // 要素を削除する。戻り値は削除した要素数。
  cout << c << endl;
  c = set2.erase("python");
  cout << c << endl;

  auto c2 = set2.find("array"); // 要素を探してイテレータを返す
  if(c2 == set2.end()) { // 要素が見つかったかどうかの判定 find で判定
    cout << "array is not found." << endl;
  } else {
    cout << "array is found." << endl;
  }
  if(set2.count("array") == 0) { // 要素が見つかったかどうかの判定 count(要素数)で判定
    cout << "array is not found." << endl;
  } else {
    cout << "array is found." << endl;
  }
  cout << distance(set2.begin(), c2) << endl; // 見つけた要素の添え字を表示する 

  auto c3 = set2.find("python"); // 要素を探してイテレータを返す
  if(c3 == set2.end()) { // 要素が見つかったかどうかの判定
    cout << "python is not found." << endl;
  } else {
    cout << "python is found." << endl;
  }
  if(set2.count("python") == 0) { // 要素が見つかったかどうかの判定 count(要素数)で判定
    cout << "python is not found." << endl;
  } else {
    cout << "python is found." << endl;
  }
  cout << distance(set2.begin(), c3) << endl; // 見つからなかった場合は添え字が集合のサイズになる 

  // map と同等の処理
  vector<int> items{1, 2, 3};
  vector<int> mapped(items.size()); // items と同じ要素数で初期化する
  transform(items.begin(), items.end(), mapped.begin(), square);
  for(auto& i: mapped) {
    cout << i << endl;
  }

  transform(items.begin(), items.end(), mapped.begin(), [](int x){return x * x;}); // 無名関数を使った場合
  for(auto& i: mapped) {
    cout << i << endl;
  }

  // filter と同等の処理
  vector<int> filtered;
  copy_if(items.begin(), items.end(), back_inserter(filtered), isOdd);
  for(auto& i: filtered) {
    cout << i << endl;
  }

  copy_if(items.begin(), items.end(), back_inserter(filtered), [](int x){return x%2!=0;}); // 無名関数を使った場合
  for(auto& i: filtered) {
    cout << i << endl;
  }

  // reduce と同等の処理
  int result = accumulate(items.begin(), items.end(), (int) 0, add);
  cout << result << endl;

  result = accumulate(items.begin(), items.end(), (int) 0, [](int a, int b){return a + b;}); // 無名関数を使った場合
  cout << result << endl;

  // 残念ながらこれらを使っても二乗和など配列に対する様々な操作を簡潔に(1行で)書くことはできない

  return 0;
}
